Add Actor-based RPC handler

Bluexin opened the pull request:
Looking for feedback :)

Extension functions could be added to wrap `send`/`offer` calls, but wanted to get feedback from people who might be more experienced with actors and coroutines in general before more work goes into this.

From my testing it works as expected. I didn't get any of the issues described in #5.

Bluexin committed:
Add experimental Actor-based RPC handler

Signed-off-by: bluexin <bluexin.games@gmail.com>

Bluexin commented:
Other ways for doing this which I considered were : 
- Have the actor have a receive and a send channel, with the receive channel working like here but the send channel being used to send updates to the consumer, which means they'll have to take care of digesting the output
- *I can't remember, maybe it'll come back after a good nap*

xeruf commented:
Two thoughts:
-  I never really liked the way you designed the `onReady`. If I set onReady and then called ifConnectedOrLater, the latter would override the former. It should have some kind of queue-behaviour, so you add instead of set
- Why wrap the callbacks in classes to make them RPCMessages? You could make them interfaces that implement RPCMessage or something like that. These ~20 lines look really WET to me.

Bluexin commented:
You'll notice I didn't add `ifConnectedOrLater` here (mainly for that reason). However I could make all of the callbacks be more like registering (aka store a list instead of only the latest one), if that's what you mean.

I'm not sure how to do that while keeping the actor model (which means communicating trough messages which will be read on the actor's thread/coroutine). I could add extension methods to send the message automatically, ie : 
```kotlin
fun SendChannel<RPCMessage>.onReady(callback: ReadyCallback) =
    this.offer(SetOnReady(callback))

suspend fun SendChannel<RPCMessage>.onReadyS(callback: ReadyCallback) =
    this.send(SetOnReady(callback))


val a = rpcActor()
// Not suspending, shouldn't be an issue anyway because our actor
// has unlimited mailbox
a.onReady {
    logger.log("Ready!")
}
// Suspending call
a.onReadyS {
    logger.log("Ready!")
}
a.close()
```

xeruf commented:
Also, `onReceive` shouldn't be suspend

Bluexin commented:
It suspends with the Connect message

Also, maybe waiting for Kotlin/kotlinx.coroutines#485 might be a better idea, not sure.

xeruf commented:
When you send a message, you don't expect to wait until it's executed, you just send it and expect it to be handled.
It shouldn't be suspend if _a single message suspends_, that is extremely confusing. Use launch instead of with, suspension solved.

Bluexin committed:
Add a second actor prototype for experimentation

Bluexin commented:
I don't understand, you don't wait until it's executed with this.
`onReceive` isn't meant for public usage, maybe I should've set visibility properly.

However, you suspend with `send` because you wait until it's successfully sent. If you don't want to suspend there, you can use `offer` which won't suspend but doesn't offer guarantee of the message being sent (however because the channel is unlimited this shouldn't matter) -- needs more testing.

Does the new actor protorype match more with what you were thinking?

Bluexin committed:
Call disconnection callback upon channel closing

xeruf commented:
Ah okay, now I get your idea. There is a function that synchronously executes the message, so you can be sure that it has been executed when the function returns, that's why it suspends. Yeah, settings visibility modifiers really helps.
I don't think `RPCActor2` is better, if anything it looks even more WET ^^

Bluexin commented:
I definitely agree, it looks pretty WET.
I think it comes from having everything doubled for suspending and non suspending calls.
I am not sure whether it's necessary to have both, removing one or the other would reduce loc count by over 40, however it would reduce consumer freedom. I am not aware of a cleaner way to keep freedom without having this much of a bloat.

I am also not sure whether we need to keep both suspending and non-suspending calls, as unlimited mailbox channel *should* make it pretty much safe to use non-suspending `offer` on (as in, it should never fail to add a message). However, I don't think there's any drawbacks to using the suspending call when in a suspending context anyway.

Bluexin commented:
I'll also add that I preferred using the second one in the examples, but that's probably because of the helper methods, which could be added to the first actor as well, but would make it even lengthier.

Bluexin committed:
Run code inspection

xeruf commented:
`RPCActor` doesn't even have a `send` and `offer` function, so whatever you were trying to explain to me there is invalid. If you implement that then it is easily superior to `RPCActor2`.

Bluexin commented:
Well when you build it you get the channel back, which does have send & offer.

Basically the second one wraps the channel without exposing it, while with the first one you're not getting a reference to the actor itself but only it's channel (hence why `RPCActor` is `private` while `RPCActor2` is `public`).

I'm more worried about the usage (exposed API) rather than the implementation details here.
As the second hides all implementation details (including the use of Channel), it's public API wouldn't change when the new Actors are added to the Coroutines library.
However, both could be changed to pure Kotlin implementation (without calling to natives) without changing their public API.

xeruf commented:
Ah, now I understand the concept.
> As the second hides all implementation details (including the use of Channel), it's public API wouldn't change when the new Actors are added to the Coroutines library.
Wut? Nearly everything in it is public.

The more I look at it, the more I like the first concept. It actually hides the implementation details completely, it doesn't even expose RPCActor itself.

xeruf commented:
I really like the idea of having a separate send channel and just throwing the events at the user instead of having to set callbacks for everything that can happen.

Bluexin commented:
Actually, if I were to change the implementation from Channel to whatever else (including non actor-like behaviour) might come in the future, the second's public API wouldn't change.

>Wut? Nearly everything in it is public.

If you watch closely, all the public methods have an implementation-agnostic signature.
The reason it looks to be much more than the first one is, again, all the helper methods (`act` and `actS` should be made private).
The first one exposes the messages and the channel, but not the actor worker itself.

>I really like the idea of having a separate send channel and just throwing the events at the user instead of having to set callbacks for everything that can happen.

Agreed, and that would be more actor-like. I'll try to draft a design which would try to be as implementation-agnostic as possible, and give the most flexibility to the consumer.

Basically exposing a channel to send messages to it, and giving it a channel to send messages to.

Bluexin committed:
Add a third actor-based prototype

This one is closer to the actual actor model, sending messages to a
channel instead of relying on callbacks.

Bluexin commented:
(rebased onto master to remove the commits I cherry-picked)

Bluexin committed:
Improve Actor 3 example output reading

xeruf commented on build.gradle.kts:
```diff
@@ -41,6 +41,7 @@ dependencies {
     implementation("org.slf4j", "slf4j-api")
     implementation("io.github.microutils", "kotlin-logging", prop("kotlinLoggingVersion"))
     compileOnly("org.slf4j", "slf4j-simple", prop("slf4jVersion"))
+    testRuntime("org.slf4j:slf4j-simple:${prop("slf4jVersion")}")
```
what the heck?

Bluexin replied:
This makes it so drpc4k tests properly have the slf4j implementation on runtime path, otherwise it'd complain and use NO-OP.
I checked, and this does not leak into consumer's runtime at all (the resulting pom for the library is still the same).

You might not have had the issue before if you used IDEA without using separate modules per source sets (which makes it use a single module for both main and test, resulting in dependencies leaking from one to the other).

xeruf replied:
but what about `compileOnly`? In my experience the main dependencies were also always available for testing?
Anyhow, please keep the notation consistent (commas, not colons).

Bluexin replied:
Fixed the notation, thought I did that but forgot about it.
`compileOnly` isn't enough when using separate modules per source sets (properly separating test from main dependencies)
Also, it was meant to replace `compileOnly` as we only need it to get logging when running the tests.

Bluexin commented:
Alright so this third one is closer to the actor model.
Got rid of the callbacks, and replaced them with an output channel.
Still missing a way to query whether it's currently connected, using `CompletableDeferred` like in previous prototypes was very non actor-like, however making a message which in turn sends a message to the output channel with whether it's connected feels a bit unintuitive (though from what I've read about the actor model that would be the correct way to do it)

Bluexin committed:
Fix extra useless GlobalScope launches

Bluexin commented on src/main/kotlin/be/bluexin/drpc4k/jna/RPCActor3.kt:
```diff
@@ -0,0 +1,150 @@
+            connected = false
+            GlobalScope.launch {
+                output.send(RPCOutputMessage.Errored(errorCode, message))
+            }
```
Not too sure about the usage of GlobalScope here.
Maybe making it a sub-job of the actor scope might be better (not using the actor scope itself, otherwise further processing would be blocked if the output channel is full), but would require me to store the actor scope (no idea whether good or bad).

xeruf replied:
That sounds good, since the idea of a Scope is to contain everything that belongs to one thing under a hood as far as I know, so this would be ideal.

xeruf commented:
Honestly, you don't need to query whether it's connected. Sending out a signal should suffice, the receiver can always store the state if he wants to.

Bluexin committed:
Respect dependencies style (commas vs colons)

Bluexin committed:
Remove useless dependency

Bluexin committed:
Add scope to the RPC actor to launch children in

Bluexin committed:
Fix default context and make actor lazy

Bluexin committed:
Add kdoc for the third actor prototype

Bluexin commented:
I added some documentation, however I'm not very good at English so please tell me if some parts seem weirdly phrased, are unclear or whatever.

Also, please tell me what you think of this prototype. I think this one follows the Actor model the most, and felt okay to use.

Bluexin committed:
Update libraries

Bluexin committed:
Remove refused Actor prototypes

Bluexin committed:
Rename RPCActor3 to RPCActor

xeruf commented on gradle.properties:
```diff
@@ -1,20 +1,34 @@
-#Plugins versions
-kotlin_version=1.3.0
+#
+# Copyright (c) 2019 Arnaud 'Bluexin' Sol�
```
you are not seriously adding a copright header to gradle.properties? Please...

xeruf commented on src/main/kotlin/be/bluexin/drpc4k/jna/RPCActor.kt:
```diff
@@ -0,0 +1,262 @@
+ * // Set up receiving of updates from the RPC actor
+ * launch {
+ *     for (msg in rpcOutput) with(msg) {
+ *         when (this) {
```
why `with(msg)` instead of simply `when(msg)`?

Bluexin replied:
Because msg is the only thing used in the context, so `this` can get smart casted in the message handlers, which just makes it a bit shorter.

xeruf commented on src/main/kotlin/be/bluexin/drpc4k/jna/RPCActor.kt:
```diff
@@ -0,0 +1,262 @@
+ */
+@ObsoleteCoroutinesApi
+@ExperimentalCoroutinesApi
+fun CoroutineScope.rpcActor(output: SendChannel<RPCOutputMessage>, context: CoroutineContext = EmptyCoroutineContext):
```
```suggestion
fun CoroutineScope.rpcActor(output: SendChannel<RPCOutputMessage>, context: CoroutineContext = this.coroutineContext):
```

xeruf commented on src/main/kotlin/be/bluexin/drpc4k/jna/RPCActor.kt:
```diff
@@ -0,0 +1,262 @@
+     * Update the user's Rich Presence.
+     * The presence will be cached if used before the app has connected, and automatically sent once ready.
+     *
+     * @param presence presence to update.
```
superfluous parameter documentation

xeruf commented on src/main/kotlin/be/bluexin/drpc4k/jna/RPCActor.kt:
```diff
@@ -0,0 +1,262 @@
+        private val input: ReceiveChannel<RPCInputMessage>,
+        private val output: SendChannel<RPCOutputMessage>) {
+
+    private val logger = KotlinLogging.logger { }
```
You may move this to the top of the file so that there is only one static logger instance

Bluexin replied:
Loggers have single instance per logger name, so it won't create a new one. I feel like moving it to file might add more noise to file-level?

xeruf commented on src/main/kotlin/be/bluexin/drpc4k/jna/RPCActor.kt:
```diff
@@ -0,0 +1,262 @@
+    /**
+     * Start the actor.
+     */
+    suspend operator fun invoke() {
```
No. This should definitely not be invoke!

xeruf commented on src/main/kotlin/be/bluexin/drpc4k/jna/RPCActor.kt:
```diff
@@ -0,0 +1,262 @@
+
+    private suspend fun onReceive(msg: RPCInputMessage) {
+        when (msg) {
+            is RPCInputMessage.Connect -> with(msg) { connect(clientId, autoRegister, steamId, refreshRate) }
```
Why is the `with` necessary?

Bluexin replied:
To use all the parameters from the message, just shorter and doesn't hurt readability imo

xeruf replied:
ah ok

Bluexin committed:
Fix license headers

(blame Idea)

Bluexin committed:
Remove redundant documentation

Commit:
Use correct context when creating the actor

Co-Authored-By: Bluexin <bluexin.games@gmail.com>

Bluexin commented:
Should I make the message classes `data class` for destructuring syntax or would that not really be needed?

Bluexin committed:
Rename actor's start function to be more explicit

xeruf commented:
> Should I make the message classes `data class` for destructuring syntax or would that not really be needed?

Sounds like a good idea, because that would also generate appropriate equals and hashCode methods.

Bluexin committed:
Make messages classes `data class`

Bluexin commented:
If there is anything else you spot, or an improvement you see, please tell me as I plan to merge it into master and publish it next week.

Bluexin committed:
Bump minor and prepare for merge
